### PR TITLE
🎨 Cleaning up cachette

### PR DESCRIPTION
**Pull request by louib**

### CODE REVIEW

File: src/lib/RedisCache.ts [hunk 1]
```diff
@@ -51,26 +52,26 @@ export class RedisCache extends CacheInstance {
...
     Cachette.setCacheInstance(null);
   }
 
   /**
    * The end event is emitted by the redis client when an
    * established connection has ended.
    */
-  public static endConnectionStrategy(): void {
-    Cachette.logger.warn(`Connection lost to Redis.`);
+  public endConnectionStrategy(): void {
```
Comment (erykwarren) [inline src/lib/RedisCache.ts]: Looking at the redis code, this function can take an err parameter. Add it and re-emit it. You can pass multiple args to emit().

File: src/lib/RedisCache.ts [hunk 2]
```diff
@@ -232,11 +235,11 @@ export class RedisCache extends CacheInstance {
...
     overwrite: boolean,
   ): Promise<boolean> {
-    Cachette.logger.debug(`Setting ${key} to`, value);
+    this.emit('set', key, value);
 
     if (value === undefined) {
-      Cachette.logger.warn(`Cannot set ${key} to undefined!`);
-      return;
+      this.emit('warn', `Cannot set ${key} to undefined!`);
+      return false;
```
Comment (erykwarren) [inline src/lib/RedisCache.ts]: Throw?

File: src/lib/RedisCache.ts [hunk 3]
```diff
@@ -259,14 +262,14 @@ export class RedisCache extends CacheInstance {
        * A timeout can occur if the connection was broken during
        * a value fetching. We don't want to hang forever if this is the case.
        */
-      Cachette.logger.error(`Error while fetching from the Redis cache`, error);
+      this.emit('warn', `Error while fetching value from the Redis cache ${util.inspect(error)}`);
```
Comment (erykwarren) [inline src/lib/RedisCache.ts]: error as a separate arg.

File: src/test/LocalCache_test.ts [hunk 4]
```diff
@@ -7,26 +7,32 @@ import * as Bluebird from 'bluebird';
...
 
   it('can set values', async () => {
-
     const cache = new LocalCache();
     cache.setValue('key', 'value');
     const value = await cache.getValue('key');
     expect(value).to.equal('value');
+  });
 
+  it('will not throw if we set a value of undefined', async () => {
```
Comment (erykwarren) [inline src/test/LocalCache_test.ts]: why?

File: src/test/RedisCache_test.ts [hunk 5]
```diff
@@ -52,32 +52,34 @@ describe('RedisCache', () => {
 
   });
 
-  describe('retryStrategy function', () => {
+  describe('retryStrategy', () => {
 
     it('Will not try to reconnect when it was never able to connect', () => {
```
Comment (erykwarren) [inline src/test/RedisCache_test.ts]: lowercase the first char

File: src/lib/LocalCache.ts [hunk 6]
```diff
@@ -16,12 +15,12 @@ export class LocalCache extends CacheInstance {
...
-  public setValue(key: string, value: cachableValue, ttl: number = 0, overwrite: boolean = true): Promise<boolean> {
-    Cachette.logger.debug(`Setting ${key} to`, value);
+  public async setValue(key: string, value: cachableValue, ttl: number = 0, overwrite: boolean = true): Promise<boolean> {
+    this.emit('set', key, value);
 
     if (value === undefined) {
-      Cachette.logger.warn(`Cannot set ${key} to undefined!`);
-      return;
+      this.emit('warn', `Cannot set ${key} to undefined!`);
+      return false;
```
[the remark this answers is not in the document]
Comment (louib) [reply src/lib/LocalCache.ts]: @erykwarren I'm not sure how it's used right now, so I'd rather not change it.

File: src/test/LocalCache_test.ts (hunk 4 above)
Comment (louib) [reply src/test/LocalCache_test.ts]: @erykwarren My call is that not setting the value if there's a bug in the client code is better than throwing, but I'm not sure right now.

File: src/lib/LocalCache.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (louib) [reply src/lib/LocalCache.ts]: See comment below

File: package.json [hunk 7]
```diff
@@ -1,8 +1,9 @@
 {
   "name": "cachette",
   "version": "1.0.0",
-  "description": "Cache library with easy decorator and concurrent initialization",
+  "description": "Resilient cache library supporting concurrent requests through local cache or Redis.",
```
[the remark this answers is not in the document]
Comment (louib) [reply package.json]: :blush:

File: src/test/RedisCache_test.ts [hunk 8]
```diff
@@ -54,7 +54,7 @@ describe('RedisCache', () => {
 
   describe('retryStrategy', () => {
 
-    it('Will not try to reconnect when it was never able to connect', () => {
+    it('will not try to reconnect when it was never able to connect', () => {
```
Comment (erykwarren) [inline src/test/RedisCache_test.ts]: and the other occurrences in this file

File: tsconfig.json [hunk 9]
```diff
@@ -3,17 +3,21 @@
...
     "sourceMap": true,
     "moduleResolution": "node",
     "experimentalDecorators": true,
-    "noUnusedParameters": true,
-    "noUnusedLocals": true
+    "noUnusedLocals": true,
+    "baseUrl": ".",
+    "paths": {
+      "node": ["node/v6"]
+    }
```
Comment (erykwarren) [inline tsconfig.json]: I believe this one can go. We now install a specific version of the @types/node package instead.

Comment (erykwarren) [reply tsconfig.json]: Perhaps we could experiment with stricter options that were introduced, such as
    "noImplicitAny": true,
    "strictNullChecks": true

File: src/lib/Cachette.ts [hunk 10]
```diff
@@ -72,6 +69,7 @@ export module Cachette {
    * Returns the cache instance.
    */
   export function getCacheInstance(): CacheInstance {
+
```
Comment (erykwarren) [inline src/lib/Cachette.ts]: On line 44, you could emit('hit', key)

Comment (erykwarren) [inline src/lib/Cachette.ts]: On line 60, we need 
```js
   if (result !== undefined) {
    await instance.setValue(key, result, ttl, overwrite);
  }
```

Comment (erykwarren) [inline src/lib/Cachette.ts]: Didn't you want to change the buildCacheKey for a property?

File: tsconfig.json (hunk 9 above)
Comment (louib) [reply tsconfig.json]: @erykwarren I just tried, and it's generating a lot warning when we set the cache instance to `null`, for example. I think I'll defer that to another PR.

File: src/lib/Cachette.ts (hunk 10 above)
Comment (louib) [reply src/lib/Cachette.ts]: at some point, yeah. Not in this PR though.

**Pull request merged**